command/agent/cache/lease_cache.go

```diff
@@ -0,0 +1,743 @@
+		}
+
+		if err := c.handleCacheClear(ctx, "token_accessor", accessor.(string)); err != nil {
+			return false, err
+		}
+
+	case path == vaultPathTokenRevokeOrphan:
+		jsonBody := map[string]interface{}{}
+		if err := json.Unmarshal(req.RequestBody, &jsonBody); err != nil {
+			return false, err
+		}
+		token, ok := jsonBody["token"]
+		if !ok {
+			return false, fmt.Errorf("failed to get token from request body")
+		}
+
+		// Find out all the indexes that are directly tied to the revoked token
+		indexes, err := c.db.GetByPrefix(cachememdb.IndexNameToken.String(), token.(string))
+		if err != nil {
+			return false, err
+		}
+
+		// Out of these indexes, one will be for the token itself and the rest
+		// will be for leases of this token. Cancel the contexts of all the
+		// leases and return from renewer goroutine for the token's index
+		// without cancelling the context. Cancelling the context of the
+		// token's renewer will evict all the child tokens which is not
+		// desired.
+		for _, index := range indexes {
+			if index.Lease != "" {
+				index.RenewCtxInfo.CancelFunc()
+			} else {
+				close(index.RenewCtxInfo.DoneCh)
+			}
+		}
+
+		// Clear the parent references of the revoked token
+		indexes, err = c.db.GetByPrefix(cachememdb.IndexNameTokenParent.String(), token.(string))
+		if err != nil {
+			return false, err
+		}
+		for _, index := range indexes {
+			index.TokenParent = ""
+			err = c.db.Set(index)
+			if err != nil {
+				c.logger.Error("failed to persist index", "error", err)
+				return false, err
+			}
+		}
+
+	case path == vaultPathLeaseRevoke:
+		// TODO: Should lease present in the URL itself be considered here?
+		// Get the lease from the request body
+		jsonBody := map[string]interface{}{}
+		if err := json.Unmarshal(req.RequestBody, &jsonBody); err != nil {
+			return false, err
+		}
+		leaseID, ok := jsonBody["lease_id"]
+		if !ok {
+			return false, fmt.Errorf("failed to get lease_id from request body")
+		}
+		if err := c.handleCacheClear(ctx, "lease", leaseID.(string)); err != nil {
+			return false, err
+		}
+
+	case strings.HasPrefix(path, vaultPathLeaseRevokeForce):
+		// Trim the URL path to get the request path prefix
+		prefix := strings.TrimPrefix(path, vaultPathLeaseRevokeForce)
+		// Get all the cache indexes that use the request path containing the
+		// prefix and cancel the renewer context of each.
+		indexes, err := c.db.GetByPrefix("request_path", namespace, "/v1"+prefix)
+		if err != nil {
+			return false, err
+		}
+		for _, index := range indexes {
+			index.RenewCtxInfo.CancelFunc()
+		}
+
+	case strings.HasPrefix(path, vaultPathLeaseRevokePrefix):
+		// Trim the URL path to get the request path prefix
+		prefix := strings.TrimPrefix(path, vaultPathLeaseRevokePrefix)
+		// Get all the cache indexes that use the request path containing the
+		// prefix and cancel the renewer context of each.
+		indexes, err := c.db.GetByPrefix("request_path", namespace, "/v1"+prefix)
```

Comment:
Should this be lease ID prefix, not request path?

Reply:
Same as above.